### PR TITLE
Abilities

### PR DESCRIPTION
**Pull request by acstanton515**
initial foundation for adding abilities data

### CODE REVIEW

**Comment by acstanton515** [comment]
> As for discovery I think we should start from units/buildings, and unpack the possibly related abilities from there. Each unit and building will have the following data properties that we could inspect.

I started with this early on, and while it does grab many abilities; some are non-interesting (thus filtered out civ_core), missing, or wrong; I have some of the code from that experiment.  There are approximately 350 files in abilities.

parse.ts and items.ts

```
diff --git a/src/attrib/parse.ts b/src/attrib/parse.ts
index f3f479f..fcd9b39 100644
--- a/src/attrib/parse.ts
+++ b/src/attrib/parse.ts
@@ -124,6 +124,8 @@ export async function parseItemFromAttribFile(file: string, data: any, civ: civC
           }
         }
 
+      const abilitiesExts = findExt(data, "squadexts", "sbpextensions/squad_ability_ext");
+
       const unit: Unit = {
         ...item,
         type: "unit",
@@ -133,6 +135,7 @@ export async function parseItemFromAttribFile(file: string, data: any, civ: civC
         sight: parseSight(ebpExts?.sight_ext),
         movement: parseMovement(ebpExts?.moving_ext),
         garrison: parseGarrison(ebpExts?.hold_ext),
+        abilities: parseAbilities(abilitiesExts),
       };
 
       return unit;
@@ -304,6 +307,15 @@ function parseUnique(ui_ext: any) {
   return ui_ext.is_unique_to_race || ui_ext.is_unique || ["UniqueBuildingUpgradeDataTemplate", "BuildingImprovedUpgradeDataTemplate"].includes(ui_ext.tooltip_data_template);
 }
 
+function parseAbilities(abilities_ext: any) {
+  return abilities_ext?.abilities
+    ? {
+      type: "ability",
+      abilities: abilities_ext?.abilities?.map((x) => (typeof x == "string" ? x : x.ability)).filter((x) => !x.includes("civ_core")) ?? [],
+    }
+    : undefined;
+}
+
 function parseInfluences(ui_ext: any) {
   if (ui_ext.ui_extra_infos)
     return ui_ext.ui_extra_infos?.reduce((inf, x) => {
diff --git a/src/types/items.ts b/src/types/items.ts
index 0f488fd..ee3c89f 100644
--- a/src/types/items.ts
+++ b/src/types/items.ts
@@ -84,6 +84,7 @@ export interface PhysicalItem extends Item {
 
 export interface Unit extends PhysicalItem {
   type: "unit";
+  abilities?: string;
   movement: {
     speed: number;
   };
```

Results:
```
..longbowman-2...
+      "abilities": {
+        "type": "ability",
+        "abilities": [
+          "abilities/toggle_abilities/english/longbow_set_up_camp",
+          "abilities/timed_abilities/english/longbow_rate_of_fire_ability",
+          "abilities/modal_abilities/core/military_neutralize_holy_site"
+        ]
       }

...royal-knight-2...
+      "abilities": {
+        "type": "ability",
+        "abilities": [
+          "abilities/always_on_abilities/french/lancer_charge_melee_boost_fre",
+          "abilities/modal_abilities/core/military_neutralize_holy_site",
+          "abilities/always_on_abilities/core/military_charge"
+        ]
       }

...ghulam-3...
+      },
+      "abilities": {
+        "type": "ability",
+        "abilities": [
+          "abilities/modal_abilities/core/military_neutralize_holy_site",
+          "abilities/always_on_abilities/core/military_charge"
+        ]
       }

...camel-rider-3...
+      "abilities": {
+        "type": "ability",
+        "abilities": [
+          "abilities/modal_abilities/core/military_neutralize_holy_site"
+        ]
       }

...mehter-2...
+      "abilities": {
+        "type": "ability",
+        "abilities": [
+          "abilities/modal_abilities/core/military_neutralize_holy_site",
+          "abilities/always_on_abilities/core/military_charge"
+        ]
       }
```


We can also implement discovery in run.ts which I did in my next approach, before landing where I am now on hard-coded discovery:

run.ts

```
function findAbilities(data: any) {
  if (data?.extensions?.find((x) => x.exts === "ebpextensions/ability_ext"))
    return data?.extensions?.find((x) => x.exts === "ebpextensions/ability_ext")?.abilities?.flatMap((u) => u.ability ?? u) ?? [];
  if (data?.extensions?.find((x) => x.exts === "sbpextensions/squad_ability_ext"))
    return data?.extensions?.find((x) => x.exts === "sbpextensions/squad_ability_ext")?.abilities?.flatMap((u) => u.ability ?? u) ?? [];
}
```

> Before we do so I think it is wise to see what data properties are available from the game files and which would possibly need to be added by hand.

And this is where I feel the files are useful for basically name, description, and icon.  Anything else is hit or miss.  I think I found a few abilities that implement help_text_formatter.  recharge_time and range may be fairly accurate to capture in source/abilities.

The ToV buff isnt even in source/abilities (the influence aspect of it is, but I only find ToV details in info/buff_info).

I started to move some of my notes on what data is available in these files to this [Google Sheets](https://docs.google.com/spreadsheets/d/1J5IAkNzRDziYUOwkuptuqA4wyZFORFcSpwB1O0rzKrY/edit?usp=sharing), but have not fully captured the details.


**In conclusion,**

I am open to doing it any of these ways, but currently favor hard coding 40-50 abilities.

We may use parseAbilities for certain abilities that are applied to many units like charges and  spearwall and actually make that a prop of Units (i.e. charge: true or spearwall: true)

And my apologies, I am kind of focused on abilities for units for phase 1 but definitely keeping in mind how it would extend to buildings.

**Comment by acstanton515** [comment]
> The nice thing here is that in UIs like Explorer, you can work from the item/building buttons and passive abilities that may be available and look them up.

I'd imagine you could do this the same way you do with technologies in Explorer, by looking for matches in the modifier of the ability.  I think the func in Explorer is **modifierMatches**

**Comment by acstanton515** [comment]
@robertvanhoesel - in todays commit, I have:

-  kept _Ability extends Item_ by moving some of the properties to downstream interfaces and re-ordered parse.ts to handle those properties similarly downstream (this results in lots of changes to data files as the properties appear in a different order in the JSON, and I left those out of the commit for now).  I went this route as Item does truly have many/most properties we should adopt into Ability.
- pulled the first set of data properties from the initial set of hardcoded source/attrib/instances/abilities files
- retrofitted technologyModifiers to be able to apply to abilities as well (there is a section in technology in parse.ts where you attempt to pull the properties from float_properties, but I dont think this is actually used, is it?)

You may like what you see in the abilities/all.json now.

Let me know what you don't like.
~ A

**Comment by acstanton515** [comment]
@robertvanhoesel 

i returned class handling to the way it was so abilities does not impact files of existing data sources.  keeping on going, my current struggle is with techs that unlock abilities.

how do you want to handle techs that unlock abilities in terms of identification? They may have the same baseId and/or id and throw error if not done differently.

`Error: Duplicate item id arrow-volley-4 in abilities/timed_abilities/english/longbow_rate_of_fire_ability conflicts with upgrade_ranged_longbow_arrow_volley_eng`

I could implement overrideAge workaround for ability to properly reflect the age it is available when it is being unlocked, or create some dependency between techs that unlock abilities, but so far, there is no precedence of relationship of one item to the next.

Again, abilities are lacking detail as the requirements for arrow volley ability only mention the tech.

```
                <list name="requirements">
                        <template_reference name="required" value="requirements\required_player_upgrade" List.ItemID="-1228156972">
                                <enum name="reason" value="usage_and_display" />
                                <locstring name="ui_name" value="0" />
                                <bool name="is_present" value="True" />
                                <int name="max_completed" value="100" />
                                <int name="min_completed" value="1" />
                                <instance_reference name="upgrade_name" value="upgrade\races\english\research\upgrade_ranged_longbow_arrow_volley_eng" />
                                <bool name="include_completed" value="True" />
                                <bool name="include_queued" value="False" />
                                <group name="include_pbg_parenting">
                                        <bool name="include_child_pbgs" value="False" />
                                        <bool name="include_parent_pbgs" value="False" />
                                </group>
                        </template_reference>
                </list>
```

**Comment by acstanton515** [comment]
Appears to be no ill will in refining the error check by type:


```
    if (items.has(item.id)) {
      if (items.get(item.id)!.type == item.type) {
        throw new Error(`Duplicate item id ${item.id} in ${file} conflicts with ${items.get(item.id)!.attribName}`);
      }
    }
```

Then keep separate lists of modifiers:

```
import { technologyModifiers, abilityModifiers } from "./modifiers";
```

**Comment by acstanton515** [comment]
Abilities are starting to get added quickly now that many issues are sorted.

File: src/attrib/parse.ts [hunk 1]
```diff
@@ -87,6 +99,75 @@ export async function parseItemFromAttribFile(file: string, data: any, civ: civC
...
+      //       addEffect("rangedAttack", value);
+      //       addEffect("siegeAttack", value);
+      //       addEffect("fireAttack", value);
+      //     } else if (id === "multiplier") {
+      //       // ignore
+      //     } else console.log("Unknown float property", id, value, item.attribName, item.id);
+      //   }
+      // }
+      
+      const ability: Ability = {
```
Comment (robertvanhoesel) [inline src/attrib/parse.ts:130]: I'd suggest we put ability related properties under a namespace here, allowing a bit more relaxed naming.
```js
ability = {
    ...item,
    type: 'ability',
    ability: {
      activation: '...',
      cooldown: '...'
    }
}
```

File: src/attrib/parse.ts [hunk 2]
```diff
@@ -87,6 +99,75 @@ export async function parseItemFromAttribFile(file: string, data: any, civ: civC
...
+        range: data.ability_bag.range / 4,
+        //rechargeTime: data.ability_bag.recharge_time,
+        //toggleGroup: data.ability_bag.toggle_ability_group ?? "",
+        effects,
+      };
+      delete(ability["unique"])
+      delete(ability["producedBy"])
+      
+      if (ability["activation"]=="toggle") ability["toggleGroup"]=data.ability_bag.toggle_ability_group;
+      if (!(ability["activation"]=="always")) ability["activationRechargeTime"]=data.ability_bag.recharge_time;
```
Comment (robertvanhoesel) [inline src/attrib/parse.ts:147]: Feel free to divert from game naming if it would make more sense to the average user. I.e. I'd say 'cooldown' is perfectly fine here instead.

File: src/attrib/parse.ts [hunk 3]
```diff
@@ -87,6 +99,75 @@ export async function parseItemFromAttribFile(file: string, data: any, civ: civC
...
+      
+      const ability: Ability = {
+        ...item,
+        type: "ability",
+        displayClasses: "",
+        classes: [],
+        // abilityType: file.split("/")[1].split("_")[0],
+        // activation: data.ability_bag.activation,
+        activation: file.split("/")[1].split("_")[0],
+        range: data.ability_bag.range / 4,
```
Comment (robertvanhoesel) [inline src/attrib/parse.ts:138]: Is this the casting range / influence range? Maybe something like 'aura', 'auraRange', or an other would work better?

File: src/attrib/parse.ts [hunk 4]
```diff
@@ -87,6 +99,75 @@ export async function parseItemFromAttribFile(file: string, data: any, civ: civC
...
+      delete(ability["unique"])
+      delete(ability["producedBy"])
+      
+      if (ability["activation"]=="toggle") ability["toggleGroup"]=data.ability_bag.toggle_ability_group;
+      if (!(ability["activation"]=="always")) ability["activationRechargeTime"]=data.ability_bag.recharge_time;
+      
+      return ability;
+    }
+    
+    if (type === ITEM_TYPES.ABILITIES && file.startsWith("info/buff_info")) {
```
Comment (robertvanhoesel) [inline src/attrib/parse.ts:152]: Do you think it is possible to try and rework this second implementation into a nested if statement or flag? there's quite some duplication here, but maybe I'm missing something.

I.e.
```js
const isBuff = file.startsWith("info/buff_info");
if(isBuff) {
 // some specific mapping
 }
```

File: src/types/items.ts [hunk 5]
```diff
@@ -123,13 +123,25 @@ export interface Weapon {
   pbgid?: number;
 }
 
+export interface Ability extends Item {
+  type: "ability";
+  //abilityType?: "always_on" | "timed" | "toggle" | "modal";
+  //activation?: string;
+  activation: "always" | "timed" | "toggle" | "modal";
```
Comment (robertvanhoesel) [inline src/types/items.ts:130]: Do you know what is the modal type activation? Like vizier points or something?

File: src/attrib/parse.ts [hunk 6]
```diff
@@ -63,9 +65,19 @@ export async function parseItemFromAttribFile(file: string, data: any, civ: civC
...
+      costs = parseCosts(data.ability_bag.cost_to_player, data.ability_bag.recharge_cost, 0 );
+    else if (type === ITEM_TYPES.ABILITIES && file.startsWith("info/buff_info"))
+      costs = {};
+    else
+      costs = parseCosts(ebpExts?.cost_ext?.time_cost?.cost || data.upgrade_bag?.time_cost?.cost, ebpExts?.cost_ext?.time_cost?.time_seconds || data.upgrade_bag?.time_cost?.time_seconds, ebpExts?.population_ext?.personnel_pop );
 
-    const icon = await useIcon(ui_ext.icon_name, type, id);
+    let icon;
+    if (type === ITEM_TYPES.ABILITIES && file.startsWith("info/buff_info")) 
+      icon = await useIcon(ui_ext.icon.slice(6), type, id);
```
Comment (robertvanhoesel) [inline src/attrib/parse.ts:78]: there's also icons tucked away in `ability_bag.ui_info.icon_name` for i.e. camel support.

File: src/attrib/config.ts [hunk 7]
```diff
@@ -10,8 +10,29 @@ export const ESSENCE_FOLDER = path.join(SOURCE_FOLDER, "/essence/attrib");
 
 // Unit files not discovered by sync.ts
 export const hardcodedDiscovery = {
-  rus: ["sbps/races/rus/unit_militia_2_rus", "upgrade/races/rus/units/upgrade_militia_3", "upgrade/races/rus/units/upgrade_militia_4"],
-  mongols: ["sbps/races/mongol/unit_khan_2_mon", "sbps/races/mongol/unit_khan_3_mon", "sbps/races/mongol/unit_khan_4_mon"],
+  rus: [
```
Comment (robertvanhoesel) [inline src/attrib/config.ts:13]: I'm not a super big fan of this approach, but can image it makes the process much easier. Perhaps, once finished, we can see if we can invert it (auto loading all and stripping the once we can't support)

File: src/attrib/workarounds.ts [hunk 8]
```diff
@@ -148,6 +148,140 @@ workaround("Make English Wynguard Forces available from Imperial Age", {
...
+  mutator: (item) => {
+    item.age = 4;
+    item.id = `${item.baseId}-${item.age}`;
+    item.activation = "always";
+    item.range = 0;
+    item.costs = { ...NO_COSTS, popcap:0 };
+  },
+});
+
+workaround("Fix age and add missing info for spirit_way", {
```
Comment (robertvanhoesel) [inline src/attrib/workarounds.ts:200]: It's pretty annoying the info in files like `attrib/instances/info/buff_info/races/chinese/spirit_way_chi.xml` isn't logically referenced from `abilities/always_on_abilities/chinese/spirit_way.xml` or vice versa.

Roughly looking at all the buffs I would say that maybe we can get away with auto joining in certain buff info based on (partial) name match? Or otherwise some additional configuration/mapping near the hardcoded ones. It seems the dev have some hardcoded things going in themselves. 

Especially hardcoded translations like the one below are high maintenance otherwise.

**Review by robertvanhoesel — COMMENTED** [review]
Hey great work mate! 

Really glad you are taking the time for this. Minus some nits and naming things I have not much to add, except that we should see if we can have as much as possible happening dynamically instead of static configs or workaround. If you need help with that let me know, I think I'll have some time this week.

File: src/attrib/parse.ts (hunk 4 above)
Comment (acstanton515) [reply src/attrib/parse.ts:152]: Absolutely

File: src/types/items.ts (hunk 5 above)
Comment (acstanton515) [reply src/types/items.ts:130]: the only modal I have noted is:
abilities/modal_abilities/french/deploy_pavise_fre

it may be something we abstract away from

i cant find vizier points in either places but i probably am overlooking...

```
ls abilities/*/ottoman/*
ls info/buff_info/races/ottoman/ -lh
```

modal abilities are mostly uninteresting to me and there are about 56 files:

```
ls abilities/modal_abilities/*/* -lh
```

File: src/attrib/config.ts (hunk 7 above)
Comment (acstanton515) [reply src/attrib/config.ts:13]: I understand.  I am very unhappy with it too.  I worked on a couple different dynamic approaches first before succumbing to hardcodes.  I presented evidences in comments of why it was difficult and how much filtering out we would have to do.  The data is just night and day unclean versus the other types.

```
$ ls -Rl abilities/ | grep xml | wc -l
350
$ ls -Rl info/buff_info/ | grep xml | wc -l
102
```

450 ability files and I think 40 abilities are in my first round of scope.

Happy to work with you directly on it on Discord or see what you find.

File: src/types/items.ts (hunk 5 above)
Comment (acstanton515) [reply src/types/items.ts:130]: I like it.  Furthermore, would you like to relate any of the props of abilities like the cooldown to toggle/manual?  There are possibly more properties to go after like if it works for allies or how it targets and how that related to having an aura or not.  Influence based abilities, etc.

I personally don't see vizier as abilities outside of Mehter drums.  If they are handled in techs, it should be sufficient.

File: src/attrib/parse.ts (hunk 2 above)
Comment (acstanton515) [reply src/attrib/parse.ts:147]: do you have a suggestion for toggleGroup?

File: src/attrib/parse.ts (hunk 1 above)
Comment (acstanton515) [reply src/attrib/parse.ts:130]: I'm not sure on this one.  More traversal complexity.  Could potentially push the unique traits down into the effects.

File: src/attrib/config.ts (hunk 7 above)
Comment (acstanton515) [reply src/attrib/config.ts:13]: I'll take another shot at it.

File: abilities/abbasid/house-of-wisdom-influence-1.json [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    "total": 0,
+    "popcap": 0
+  },
+  "producedBy": [
+    "house-of-wisdom"
+  ],
+  "icon": "https://data.aoe4world.com/images/abilities/house-of-wisdom-influence-1.png",
+  "active": "always",
+  "auraRange": 0,
+  "effects": []
```
Comment (robertvanhoesel) [inline abilities/abbasid/house-of-wisdom-influence-1.json:29]: Add workaround for fire armor effect to buildings.

File: abilities/chinese/great-wall-buff-chi-4.json [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  },
+  "producedBy": [],
+  "icon": "https://data.aoe4world.com/images/abilities/great-wall-buff-chi-1.png",
+  "active": "always",
+  "auraRange": 0,
+  "effects": [
+    {
+      "property": "rangedAttack",
+      "select": {
+        "id": [
```
Comment (robertvanhoesel) [inline abilities/chinese/great-wall-buff-chi-4.json:32]: Could also go for `class: [['ranged','infantry']]` so if at some point Chinese units would get added they are also targeted

File: abilities/chinese/imperial-spies-1.json [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  "baseId": "imperial-spies",
+  "type": "ability",
+  "name": "Imperial Spies",
+  "pbgid": 206838,
+  "attribName": "imperial_spies",
+  "age": 1,
+  "civs": [
+    "ch"
+  ],
+  "description": "Reveal location of enemy workers for 10 seconds.",
```
Comment (robertvanhoesel) [inline abilities/chinese/imperial-spies-1.json:12]: Maybe add effect 'unknown' with the duration

File: abilities/delhi/tower-of-victory-aura-2.json [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+      "select": {
+        "class": [
+          [
+            "infantry"
+          ]
+        ]
+      },
+      "effect": "multiply",
+      "value": 0.83,
+      "type": "ability"
```
Comment (robertvanhoesel) [inline abilities/delhi/tower-of-victory-aura-2.json:39]: I would argue this is passive or influence, as it is a permanent effect

File: abilities/english/abbey-healing-1.json [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    "total": 0,
+    "popcap": 0
+  },
+  "producedBy": [
+    "abbey-of-kings"
+  ],
+  "icon": "https://data.aoe4world.com/images/abilities/abbey-healing-1.png",
+  "active": "always",
+  "auraRange": 7.5,
+  "effects": []
```
Comment (robertvanhoesel) [inline abilities/english/abbey-healing-1.json:29]: We have an effect / modifiable property 'healingRate'

File: abilities/english/mill-influence-1.json [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    "total": 0,
+    "popcap": 0
+  },
+  "producedBy": [
+    "mill"
+  ],
+  "icon": "https://data.aoe4world.com/images/abilities/mill-influence-1.png",
+  "active": "always",
+  "auraRange": 0,
+  "effects": []
```
Comment (robertvanhoesel) [inline abilities/english/mill-influence-1.json:29]: Property foodGatherRate

File: abilities/english/network-of-castles-1.json [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+          ],
+          [
+            "ranged"
+          ],
+          [
+            "siege"
+          ]
+        ],
+        "id": [
+          "villager"
```
Comment (robertvanhoesel) [inline abilities/english/network-of-castles-1.json:52]: Good one!

File: abilities/hre/emergency-repairs-1.json [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+  "displayClasses": "",
+  "costs": {
+    "food": 0,
+    "wood": 0,
+    "stone": 0,
+    "gold": 0,
+    "total": 0,
+    "popcap": 0
+  },
+  "producedBy": [
```
Comment (robertvanhoesel) [inline abilities/hre/emergency-repairs-1.json:23]: Hmmm, don't we have class 'building' ? Maybe we should haha

File: abilities/hre/relic-garrisoned-keep-3.json [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+  "producedBy": [
+    "outpost",
+    "stone-wall-tower",
+    "keep",
+    "elzbach-palace"
+  ],
+  "icon": "https://data.aoe4world.com/images/abilities/relic-garrisoned-1.png",
+  "active": "always",
+  "auraRange": 0,
+  "effects": []
```
Comment (robertvanhoesel) [inline abilities/hre/relic-garrisoned-keep-3.json:32]: Effects missing

**Review by robertvanhoesel — COMMENTED** [review]
Looks very good! Nice work :)

**Comment by acstanton515** [comment]
@robertvanhoesel are you comfortable with the file name change for technologies.ts to modifiers.ts?  Would you be able to resolve the conflict when you are ready to pull?

**Comment by robertvanhoesel** [comment]
> @robertvanhoesel are you comfortable with the file name change for technologies.ts to modifiers.ts? Would you be able to resolve the conflict when you are ready to pull?

If it's not too much, maybe best to create a folder 'modifiers' with index that merges `abilities.ts` and `technologies.ts`? I agree it needs a bit of restructure, just hoping to make it easy for ourselves to navigte

**Comment by acstanton515** [comment]
do we want to delete classes property of abilities; currently all empty:

`data (abilities) $ grep classes abilities/all.json 
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],
      "classes": [],`

**Comment by acstanton515** [comment]
> > @robertvanhoesel are you comfortable with the file name change for technologies.ts to modifiers.ts? Would you be able to resolve the conflict when you are ready to pull?
> 
> If it's not too much, maybe best to create a folder 'modifiers' with index that merges `abilities.ts` and `technologies.ts`? I agree it needs a bit of restructure, just hoping to make it easy for ourselves to navigte

what do you mean an index that merges? they both benefit from access to the 10 const objects and fns; so those would have to  be exported

**Comment by robertvanhoesel** [comment]
> > > @robertvanhoesel are you comfortable with the file name change for technologies.ts to modifiers.ts? Would you be able to resolve the conflict when you are ready to pull?
> > 
> > 
> > If it's not too much, maybe best to create a folder 'modifiers' with index that merges `abilities.ts` and `technologies.ts`? I agree it needs a bit of restructure, just hoping to make it easy for ourselves to navigte
> 
> what do you mean an index that merges? they both benefit from access to the 10 const objects and fns; so those would have to be exported

In that case (you are right) probably best to have:
`/modifiers/index.ts` if needed at all to merge and re-export all effects
`/modifiers/config.ts` that contains the const and helpers
`/modifiers/abilities` and `/modifiers/technologies`

> do we want to delete classes property of abilities; currently all empty:

What do you think of setting them to `['Passive', 'Ability']` and `['Influence']` kinda classes?

File: abilities/delhi/tower-of-victory-aura-2.json (hunk 12 above)
Comment (acstanton515) [reply abilities/delhi/tower-of-victory-aura-2.json:39]: Below is what I have so far on this topic:

So far, every ability has **"type":"ability"** and every abilities effects/modifiers has **"type":"ability"**

ability has an **active** property with three possible values:

- always
- manual
- toggle

every ability with **"active":"manual"** should and currently does have a **duration** in the effects


some ability with **"active":"always"** has **duration**:

- inspired-warriors-3
- lancer-charge-bonus-damage-2
- spirit-way-4
- huntress-stealth-4
- saints-blessing-3

Recall that **"active"** is obtained directly from the **/abilities** file path or a workaround if in **/info/buff_info**

File: abilities/hre/emergency-repairs-1.json (hunk 16 above)
Comment (acstanton515) [reply abilities/hre/emergency-repairs-1.json:23]: producedBy should be town-center as influence of TC grants ER.
effects should have healingRate apply to class buildings

Thoughts?

File: abilities/hre/relic-garrisoned-keep-3.json (hunk 17 above)
Comment (acstanton515) [reply abilities/hre/relic-garrisoned-keep-3.json:32]: Honestly, I have not done any effects of this next round of abilities discovered by this latest commit.  Will work on that soon.

**Comment by acstanton515** [comment]
> > > > @robertvanhoesel are you comfortable with the file name change for technologies.ts to modifiers.ts? Would you be able to resolve the conflict when you are ready to pull?
> > > 
> > > 
> > > If it's not too much, maybe best to create a folder 'modifiers' with index that merges `abilities.ts` and `technologies.ts`? I agree it needs a bit of restructure, just hoping to make it easy for ourselves to navigte
> > 
> > 
> > what do you mean an index that merges? they both benefit from access to the 10 const objects and fns; so those would have to be exported
> 
> In that case (you are right) probably best to have: `/modifiers/index.ts` if needed at all to merge and re-export all effects `/modifiers/config.ts` that contains the const and helpers `/modifiers/abilities` and `/modifiers/technologies`
> 
> > do we want to delete classes property of abilities; currently all empty:
> 
> What do you think of setting them to `['Passive', 'Ability']` and `['Influence']` kinda classes?

I'm struggling to understand reasoning for creating so many files for modifiers.  I think the complexity is low with having it in a single file and the intent of the file is still clear.  I am comfortable with the way it is.  I'm happy to do this modification if you can show me what you would do in index.ts.


As to classing the abilities, I'd be happy to convert the "active" type into classes.  Every ability has a class of ability and a class of the value we have in "active" today.  Still more work to do in categorizing these abilities I think, in line with the other discussion we have on ToV.

Plus, some other topics to post.

**Comment by acstanton515** [comment]
There are maybe a dozen abilities that are unlocked by a tech (think: forced-march).  Do we want to modify the tech effect to say **"property":"unlockAbility"**, and then have the true effect in the abilities? Or do we exclude the "tech-ability" from abilities and keep all its data and ability data inside the tech?

There are some techs that upgrade abilities.  How do we want to handle that?  The tech could have  **"property":"upgradeAbility"** (e.g. royal knight charge upgrade, network of citadels, khan whistling arrow)

**Comment by acstanton515** [comment]
There are some abilities that change with age.  How do we want to handle that? (french vill production boost, english mill and farm rate, etc.)

**Comment by acstanton515** [comment]
I'm starting to look ahead at charge, spearwall, and secondary weapons.  I'm thinking instead of charge and spearwall as an ability, to actually represent it as a property of unit.  Logic will have to tie the secondary weapons to whether the unit can charge and/or has a spearwall.  Just want to validate whether we should directionally add it to abilities or wait on investigating secondary weapons.

**Comment by acstanton515** [comment]
Don't know if you saw my comment about auto discovering abilities having some issues in the code:

```
//issue in these functions in that if the item was found with sbp then ebp wont discover, and vise versa
function findEbpAbilities(data: any) {
  const foundAbilities = data?.extensions?.find((x) => x.exts === "ebpextensions/ability_ext")?.abilities?.flatMap((u) => u.ability ?? u) ?? [];
  return foundAbilities;
}

function findSbpAbilities(data: any) {
  const foundAbilities = data?.extensions?.find((x) => x.exts === "sbpextensions/squad_ability_ext")?.abilities?.flatMap((u) => u.ability ?? u) ?? [];
  return foundAbilities;
}
```

File: abilities/hre/emergency-repairs-1.json (hunk 16 above)
Comment (robertvanhoesel) [reply abilities/hre/emergency-repairs-1.json:23]: I'm really thinking this in terms of 'in game, in the UI, where do I click the button to make/trigger it'. Being in range of the TC would be a requirement, just like 'targeting a unit ' for certain abilities is a requirement.

Comment (acstanton515) [reply abilities/hre/emergency-repairs-1.json:23]: I see what you are saying.  Do you prefer "building" or "town-center"?
image
image

File: abilities/hre/relic-garrisoned-keep-3.json (hunk 17 above)
Comment (acstanton515) [reply abilities/hre/relic-garrisoned-keep-3.json:32]: @robertvanhoesel this one has an issue adopting the modifiers because relic_dock_bonus_hre and relic_tower_keep_bonus_hre have the same baseId when modifiers are added; before I change the ID in a workaround.  

I think attribName is safer to call in modifier.  I can change the 50 abilities to reference attribName instead.

File: src/attrib/workarounds.ts (hunk 8 above)
Comment (acstanton515) [reply src/attrib/workarounds.ts:254]: There are so few info/buff_infos we resort to that I think the static method we have suffices.

File: src/attrib/config.ts (hunk 7 above)
Comment (acstanton515) [reply src/attrib/config.ts:13]: We did it!

File: abilities/hre/relic-garrisoned-keep-3.json (hunk 17 above)
Comment (robertvanhoesel) [reply abilities/hre/relic-garrisoned-keep-3.json:32]: Hmm, using the attrib name isn't a bad idea but may make it a bit harder to navigate the files. Since some of the attrib names are chosen pretty poorly. I'd recommend also including the title or base ID in the comments when switching to attrib name

File: abilities/hre/emergency-repairs-1.json (hunk 16 above)
Comment (robertvanhoesel) [reply abilities/hre/emergency-repairs-1.json:23]: Hmm this one is extremely edge-casy, but let's keep all the building in the `producedBy` like it currently is.

File: abilities/delhi/tower-of-victory-aura-2.json (hunk 12 above)
Comment (robertvanhoesel) [reply abilities/delhi/tower-of-victory-aura-2.json:39]: Oke, clear reasoning. LGTM as it is

File: src/attrib/parse.ts (hunk 1 above)
Comment (robertvanhoesel) [reply src/attrib/parse.ts:137]: > More traversal complexity

You mean this in the usage of the data?
